### ISSUE TITLE
can jsf work with schema no provider?

### ISSUE THREAD

**Issue by zzsuki** [ghandic/jsf#43]
# Schema I Used

```json
{
    "title": "AlertSync",
    "description": "\u5ba1\u8ba1\u544a\u8b66model",
    "type": "object",
    "properties": {
        "audit_label": {
            "title": "Audit Label",
            "type": "string",
            "format": "ipv4"
        },
        "category": {
            "title": "Category",
            "minimum": 1,
            "maximum": 15,
            "type": "integer"
        },
        "level": {
            "title": "Level",
            "minimum": 0,
            "maximum": 3,
            "type": "integer"
        },
        "src_mac": {
            "title": "Src Mac",
            "default": "00:00:00:00:00:00",
            "pattern": "^([0-9A-F]{2})(\\:[0-9A-F]{2}){5}$",
            "type": "string"
        },
        "src_ip": {
            "title": "Src Ip",
            "type": "string",
            "format": "ipv4"
        },
        "src_port": {
            "title": "Src Port",
            "minimum": 1,
            "maximum": 65535,
            "type": "integer"
        },
        "dst_mac": {
            "title": "Dst Mac",
            "default": "FF:FF:FF:FF:FF:FF",
            "pattern": "^([0-9A-F]{2})(\\:[0-9A-F]{2}){5}$",
            "type": "string"
        },
        "dst_ip": {
            "title": "Dst Ip",
            "type": "string",
            "format": "ipv4"
        },
        "dst_port": {
            "title": "Dst Port",
            "minimum": 1,
            "maximum": 65535,
            "type": "integer"
        },
        "l4_protocol": {
            "$ref": "#/definitions/L4ProtocolEnum"
        },
        "protocol": {
            "$ref": "#/definitions/ProtocolEnum"
        },
        "illegal_ip": {
            "title": "Illegal Ip",
            "default": [],
            "type": "array",
            "items": {
                "type": "string",
                "format": "ipv4"
            }
        },
        "last_at": {
            "title": "Last At",
            "default": "2022-12-30T14:08:30.753677",
            "type": "string",
            "format": "date-time"
        },
        "count": {
            "title": "Count",
            "default": 1,
            "minimum": 1,
            "maximum": 100000,
            "type": "integer"
        },
        "other_info": {
            "title": "Other Info",
            "type": "object"
        },
        "payload": {
            "title": "Payload",
            "pattern": "^([0-9A-F]{2})+$",
            "type": "string"
        }
    },
    "required": [
        "audit_label",
        "category",
        "level",
        "l4_protocol",
        "protocol"
    ],
    "definitions": {
        "L4ProtocolEnum": {
            "title": "L4ProtocolEnum",
            "description": "An enumeration.",
            "enum": [
                "TCP",
                "UDP"
            ],
            "type": "string"
        },
        "ProtocolEnum": {
            "title": "ProtocolEnum",
            "description": "An enumeration.",
            "enum": [
                "S7COMM",
                "MODBUS"
            ],
            "type": "string"
        }
    }
}
```

# error message

```python
Traceback (most recent call last):
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/object.py", line 40, in generate
    return super().generate(context)
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/base.py", line 49, in generate
    raise ProviderNotSetException()
jsf.schema_types.base.ProviderNotSetException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/object.py", line 40, in generate
    return super().generate(context)
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/base.py", line 49, in generate
    raise ProviderNotSetException()
jsf.schema_types.base.ProviderNotSetException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/repos/sa-data-perf/debug.py", line 14, in <module>
    print(jsf.generate())
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/parser.py", line 137, in generate
    return self.root.generate(context=self.context)
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/object.py", line 42, in generate
    return {o.name: o.generate(context) for o in self.properties if self.should_keep(o.name)}
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/object.py", line 42, in <dictcomp>
    return {o.name: o.generate(context) for o in self.properties if self.should_keep(o.name)}
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/object.py", line 42, in generate
    return {o.name: o.generate(context) for o in self.properties if self.should_keep(o.name)}
TypeError: 'NoneType' object is not iterable
```

# My question

can jsf work with schema like given, this schema was generated by `pydantic`, i'm not sure which part cause this error, hope log more specifically to tell me which property cause this error

**Comment by ghandic** [comment]
Can you attach the code for the pydantic model too? I can take a look in a few days

Sent from my iPhone

On 30 Dec 2022, at 14:16, zzsuki ***@***.***> wrote:

﻿

Schema I Used

{
    "title": "AlertSync",
    "description": "\u5ba1\u8ba1\u544a\u8b66model",
    "type": "object",
    "properties": {
        "audit_label": {
            "title": "Audit Label",
            "type": "string",
            "format": "ipv4"
        },
        "category": {
            "title": "Category",
            "minimum": 1,
            "maximum": 15,
            "type": "integer"
        },
        "level": {
            "title": "Level",
            "minimum": 0,
            "maximum": 3,
            "type": "integer"
        },
        "src_mac": {
            "title": "Src Mac",
            "default": "00:00:00:00:00:00",
            "pattern": "^([0-9A-F]{2})(\\:[0-9A-F]{2}){5}$",
            "type": "string"
        },
        "src_ip": {
            "title": "Src Ip",
            "type": "string",
            "format": "ipv4"
        },
        "src_port": {
            "title": "Src Port",
            "minimum": 1,
            "maximum": 65535,
            "type": "integer"
        },
        "dst_mac": {
            "title": "Dst Mac",
            "default": "FF:FF:FF:FF:FF:FF",
            "pattern": "^([0-9A-F]{2})(\\:[0-9A-F]{2}){5}$",
            "type": "string"
        },
        "dst_ip": {
            "title": "Dst Ip",
            "type": "string",
            "format": "ipv4"
        },
        "dst_port": {
            "title": "Dst Port",
            "minimum": 1,
            "maximum": 65535,
            "type": "integer"
        },
        "l4_protocol": {
            "$ref": "#/definitions/L4ProtocolEnum"
        },
        "protocol": {
            "$ref": "#/definitions/ProtocolEnum"
        },
        "illegal_ip": {
            "title": "Illegal Ip",
            "default": [],
            "type": "array",
            "items": {
                "type": "string",
                "format": "ipv4"
            }
        },
        "last_at": {
            "title": "Last At",
            "default": "2022-12-30T14:08:30.753677",
            "type": "string",
            "format": "date-time"
        },
        "count": {
            "title": "Count",
            "default": 1,
            "minimum": 1,
            "maximum": 100000,
            "type": "integer"
        },
        "other_info": {
            "title": "Other Info",
            "type": "object"
        },
        "payload": {
            "title": "Payload",
            "pattern": "^([0-9A-F]{2})+$",
            "type": "string"
        }
    },
    "required": [
        "audit_label",
        "category",
        "level",
        "l4_protocol",
        "protocol"
    ],
    "definitions": {
        "L4ProtocolEnum": {
            "title": "L4ProtocolEnum",
            "description": "An enumeration.",
            "enum": [
                "TCP",
                "UDP"
            ],
            "type": "string"
        },
        "ProtocolEnum": {
            "title": "ProtocolEnum",
            "description": "An enumeration.",
            "enum": [
                "S7COMM",
                "MODBUS"
            ],
            "type": "string"
        }
    }
}

error message

Traceback (most recent call last):
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/object.py", line 40, in generate
    return super().generate(context)
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/base.py", line 49, in generate
    raise ProviderNotSetException()
jsf.schema_types.base.ProviderNotSetException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/object.py", line 40, in generate
    return super().generate(context)
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/base.py", line 49, in generate
    raise ProviderNotSetException()
jsf.schema_types.base.ProviderNotSetException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/repos/sa-data-perf/debug.py", line 14, in <module>
    print(jsf.generate())
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/parser.py", line 137, in generate
    return self.root.generate(context=self.context)
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/object.py", line 42, in generate
    return {o.name: o.generate(context) for o in self.properties if self.should_keep(o.name)}
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/object.py", line 42, in <dictcomp>
    return {o.name: o.generate(context) for o in self.properties if self.should_keep(o.name)}
  File "/root/repos/sa-data-perf/venv/lib/python3.10/site-packages/jsf/schema_types/object.py", line 42, in generate
    return {o.name: o.generate(context) for o in self.properties if self.should_keep(o.name)}
TypeError: 'NoneType' object is not iterable

My question

can jsf work with schema like given, this schema was generated by pydantic, i'm not sure which part cause this error, hope log more specifically to tell me which property cause this error

**Comment by ghandic** [comment]
Problem was an object with no defined properties - a bug none the less :) fixed in https://github.com/ghandic/jsf/pull/44

**Comment by ghandic** [comment]
Released as part of 0.5.0 https://pypi.org/project/jsf/0.5.0/
